File: C_src/Easy/104_Maximum_Depth_of_Binary_Tree/MaximumDepthOfBinaryTree_104.c
#include <stdio.h>
#include <stdlib.h>

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
};

// Use DFS:
// Time Complexity: O(n)
// Space Complexity:O(logn)
// Runtime: 12 ms, faster than 10.55%
// Memory Usage: 9.1 MB, less than 100.00%
#define MAX(a, b) ((a) > (b) ? (a) : (b))

int maxDepth(struct TreeNode *root)
{
    if (!root)
    {
        return 0;
    }

    int left = maxDepth(root->left);
    int right = maxDepth(root->right);

    return MAX(left, right) + 1;
}

File: Golang_src/easy/maximumDepthOfBinaryTree_104.go
package easy

// Use recursive:
// Time Complexity: O(n)
// Space Complexity:O(logn)
// Runtime: 4 ms, faster than 92.48%
// Memory Usage: 4.4 MB, less than 15.52%
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	return 1 + max(maxDepth(root.Left), maxDepth(root.Right))
}
